src/libexpr/eval.cc

```diff
@@ -2418,6 +2433,216 @@ SingleDerivedPath EvalState::coerceToSingleDerivedPath(const PosIdx pos, Value &
 }
 
 
+
+// NOTE: This implementation must match eqValues!
+// We accept this burden because informative error messages for
+// `assert a == b; x` are critical for our users' testing UX.
+void EvalState::assertEqValues(Value & v1, Value & v2, const PosIdx pos, std::string_view errorCtx)
+{
+    // This implementation must match eqValues.
+    forceValue(v1, pos);
+    forceValue(v2, pos);
+
+    if (&v1 == &v2)
+        return;
+
+    // Special case type-compatibility between float and int
+    if ((v1.type() == nInt || v1.type() == nFloat) && (v2.type() == nInt || v2.type() == nFloat)) {
+        if (eqValues(v1, v2, pos, errorCtx)) {
+            return;
+        } else {
+            error<AssertionError>(
+                "%s with value '%s' is not equal to %s with value '%s'",
+                showType(v1),
+                ValuePrinter(*this, v1, errorPrintOptions),
+                showType(v2),
+                ValuePrinter(*this, v2, errorPrintOptions))
+                .debugThrow();
+        }
+    }
+
+    if (v1.type() != v2.type()) {
+        error<AssertionError>(
+            "%s of value '%s' is not equal to %s of value '%s'",
+            showType(v1),
+            ValuePrinter(*this, v1, errorPrintOptions),
+            showType(v2),
+            ValuePrinter(*this, v2, errorPrintOptions))
+            .debugThrow();
+    }
+
+    switch (v1.type()) {
+    case nInt:
```

Comment:
Is there a way to do enforce an exhaustive c++ switch case that will fail to compile when there cases missing? Rust has this.

Reply:
Good question, but we already enforce this with
```meson
  '-Werror=switch',
  '-Werror=switch-enum',
```

This makes the error message very hypothetical. We'd have to undo those flags, or `forceValue` needs to silently fail, or we'd have to read garbage, or some other chaotic thing we don't need to consider.

Reply:
Could a default case shadow a potential error message and should be therefore removed?

Reply:
The flags already prevent this despite the `default`, so the `default` is a bit paranoid.

I've changed the exceptions to crashes to aid debugging, and I've documented that they're unreachable.